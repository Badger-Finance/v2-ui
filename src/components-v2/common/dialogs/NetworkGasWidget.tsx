import React, { useContext, useState } from 'react';
import {
	Button,
	Dialog,
	DialogContent,
	DialogTitle,
	Grid,
	IconButton,
	Typography,
	ListItem,
	List,
} from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import { GasSpeed, Network, NetworkConfig } from '@badger-dao/sdk';
import CloseIcon from '@material-ui/icons/Close';
import clsx from 'clsx';
import { observer } from 'mobx-react-lite';
import { defaultNetwork, supportedNetworks } from '../../../config/networks.config';
import { StoreContext } from '../../../mobx/store-context';
import { Loader } from '../../../components/Loader';

const useStyles = makeStyles((theme) => ({
	title: {
		padding: theme.spacing(3, 3, 0, 3),
	},
	content: {
		padding: theme.spacing(2, 3, 3, 3),
	},
	closeButton: {
		position: 'absolute',
		right: 8,
		top: 8,
	},
	networkButton: {
		minWidth: 37,
		height: 36,
		textAlign: 'center',
	},
	loadingGas: {
		textAlign: 'center',
		width: '33%',
	},
	selectedOption: {
		border: `2px solid ${theme.palette.primary.main}`,
		color: theme.palette.primary.main,
	},
	nonSelectedOption: {
		border: '0.5px solid #848484',
	},
	networkListIcon: {
		width: 17,
		height: 17,
		marginRight: theme.spacing(1),
	},
	selectedNetworkItem: {
		width: 14,
	},
	selectedNetworkIcon: {
		width: 15,
	},
	option: {
		borderRadius: 8,
	},
	confirmButton: {
		marginTop: theme.spacing(4),
	},
	gasSection: {
		marginTop: theme.spacing(2),
	},
	gasList: {
		marginTop: theme.spacing(1),
	},
	networkOption: {
		marginTop: theme.spacing(1),
	},
}));

const networkIcons: Record<Network, string> = {
	[Network.Ethereum]: 'ethereum-network.svg',
	[Network.BinanceSmartChain]: 'bsc-network.svg',
	[Network.Arbitrum]: 'arbitrum-network.svg',
	[Network.Polygon]: 'matic-network.svg',
	[Network.xDai]: 'xdai-network.svg',
	[Network.Avalanche]: 'avalanche-network.svg',
	[Network.Fantom]: 'fantom-network.svg',
};

const NetworkGasWidget = (): JSX.Element => {
	const {
		gasPrices,
		uiState: { gasPrice, setGasPrice },
		network: networkStore,
	} = useContext(StoreContext);

	const [selectedNetwork, setSelectedNetwork] = useState(defaultNetwork);
	const [selectedGas, setSelectedGas] = useState(gasPrice);
	const [isDialogOpen, setIsDialogOpen] = useState(false);
	const classes = useStyles();

	const gasOptions = gasPrices.getGasPrices(selectedNetwork.symbol);
	const { network } = networkStore;

	const toggleDialog = () => setIsDialogOpen(!isDialogOpen);

	const getOptionClassName = (selected: boolean) => (selected ? classes.selectedOption : classes.nonSelectedOption);

	const applyChanges = async () => {
		const shouldTriggerNetworkChange = selectedNetwork.symbol !== network.symbol;

		if (shouldTriggerNetworkChange) {
			const networkConfig = NetworkConfig.getConfig(selectedNetwork.symbol);
			await networkStore.setNetwork(networkConfig.id);
		}

		setGasPrice(selectedGas);
		setIsDialogOpen(false);
	};

	return (
		<>
			<Button
				classes={{ outlined: isDialogOpen ? classes.selectedOption : undefined }}
				className={classes.networkButton}
				variant="outlined"
				onClick={toggleDialog}
			>
				<img
					className={classes.selectedNetworkIcon}
					src={`assets/icons/${networkIcons[network.symbol]}`}
					alt="selected network icon"
				/>
			</Button>
			<Dialog open={isDialogOpen} fullWidth maxWidth="sm">
				<DialogTitle className={classes.title}>
					Network
					<IconButton className={classes.closeButton} onClick={toggleDialog}>
						<CloseIcon />
					</IconButton>
				</DialogTitle>
				<DialogContent className={classes.content}>
					<Grid container>
						<Grid item container direction="column">
							<List disablePadding>
								{supportedNetworks.map((supportedNetwork, index) => (
									<ListItem
										button
										className={clsx(
											classes.option,
											classes.networkOption,
											getOptionClassName(supportedNetwork.symbol === selectedNetwork.symbol),
										)}
										key={`${supportedNetwork.name}-${index}`}
										onClick={() => setSelectedNetwork(supportedNetwork)}
									>
										<img
											className={classes.networkListIcon}
											src={`assets/icons/${networkIcons[supportedNetwork.symbol]}`}
											alt={`${supportedNetwork.name} icon`}
										/>
										<Typography variant="subtitle1">{supportedNetwork.name}</Typography>
									</ListItem>
								))}
							</List>
						</Grid>
						<Grid item container className={classes.gasSection}>
							<Typography variant="subtitle1" color="textSecondary">
								GAS
							</Typography>
							{gasPrices.initialized && gasOptions ? (
								<Grid container spacing={1} className={classes.gasList}>
									{Object.entries(gasOptions).map((price) => {
										const [key, value] = price;
										const displayValue = typeof value === 'number' ? value : value.maxFeePerGas;
										return (
											<Grid item key={key}>
												<Button
													onClick={() => setSelectedGas(key as GasSpeed)}
													className={clsx(
														classes.option,
														getOptionClassName(selectedGas === key),
													)}
												>
													{displayValue ? displayValue.toFixed(0) : 10}
												</Button>
											</Grid>
										);
									})}
								</Grid>
							) : (
								<Grid container direction="column" className={classes.loadingGas}>
									<Loader size={15} />
									<Typography variant="caption">Loading Gas...</Typography>
								</Grid>
							)}
						</Grid>
						<Button
							fullWidth
							onClick={async () => await applyChanges()}
							className={classes.confirmButton}
							variant="contained"
							color="primary"
						>
							Confirm Network
						</Button>
					</Grid>
				</DialogContent>
			</Dialog>
		</>
	);
};

export default observer(NetworkGasWidget);
