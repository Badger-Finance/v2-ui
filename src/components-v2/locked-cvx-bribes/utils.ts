import BigNumber from 'bignumber.js';
import { formatWithoutExtraZeros, numberWithCommas, unscale } from '../../mobx/utils/helpers';

export function formatBalance(balance: BigNumber): string {
  return numberWithCommas(formatWithoutExtraZeros(unscale(balance, 18)));
}
