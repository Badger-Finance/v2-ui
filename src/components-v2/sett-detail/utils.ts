import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/setts/sett-charts';
import { DelaySeverity } from '../../mobx/model/setts/sett-rewards';
import { SettDetailMode } from '../../mobx/model/setts/sett-detail';

dayjs.extend(utc);

export const ChartModeTitles = {
	[ChartMode.value]: 'Sett Value',
	[ChartMode.ratio]: 'Token Ratio',
	[ChartMode.accountBalance]: 'Your Holdings',
};

export const SettModeTitles = {
	[SettDetailMode.settInformation]: 'Vault Info',
	[SettDetailMode.userInformation]: 'User Info',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};
