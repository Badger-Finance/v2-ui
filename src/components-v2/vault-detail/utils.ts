import { Vault, VaultData } from '@badger-dao/sdk';
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/vaults/vault-charts';
import { DelaySeverity } from '../../mobx/model/vaults/vault-rewards';

dayjs.extend(utc);

export const ChartModeTitles: Record<string, string> = {
	[ChartMode.Value]: 'Sett Value',
	[ChartMode.Ratio]: 'Token Ratio',
	[ChartMode.AccountBalance]: 'My Holdings',
	[ChartMode.BoostMultiplier]: 'Badger Boost',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};

export function defaultVaultBalance(vault: Vault): VaultData {
	return {
		address: vault.vaultToken,
		name: vault.name,
		symbol: vault.asset,
		pricePerFullShare: vault.pricePerFullShare,
		balance: 0,
		value: 0,
		earnedBalance: 0,
		earnedValue: 0,
		depositedBalance: 0,
		withdrawnBalance: 0,
		tokens: [],
		earnedTokens: [],
	};
}

export interface Balance {
	balance: number;
}

export function hasBalance(input?: Balance): boolean {
	if (!input) {
		return false;
	}
	return input.balance > 0;
}
