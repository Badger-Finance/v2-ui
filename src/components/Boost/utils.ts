import { makeStyles } from '@material-ui/core/styles';
import { ClassNameMap } from '@material-ui/core/styles/withStyles';
import React from 'react';

type ComparisonConfig = {
  toCompareValue: number;
  toBeComparedValue: number;
  greaterCaseColor: string;
  lessCaseColor: string;
  defaultColor: string;
};

export const getColorFromComparison = ({
  toCompareValue,
  toBeComparedValue,
  greaterCaseColor,
  lessCaseColor,
  defaultColor,
}: ComparisonConfig): string => {
  if (toCompareValue > toBeComparedValue) return greaterCaseColor;
  if (toCompareValue < toBeComparedValue) return lessCaseColor;

  return defaultColor;
};

export const useAssetInputStyles = (
  currentValue: string,
  toCompare = 0,
): ((props?: React.CSSProperties) => ClassNameMap<'assetColor'>) => {
  return makeStyles((theme) => {
    const defaultColor = currentValue ? theme.palette.text.primary : theme.palette.text.secondary;
    const fontColor = getColorFromComparison({
      toCompareValue: Number(currentValue),
      toBeComparedValue: toCompare,
      greaterCaseColor: '#74D189',
      lessCaseColor: theme.palette.error.main,
      defaultColor,
    });

    return {
      assetColor: {
        color: fontColor,
      },
    };
  });
};
