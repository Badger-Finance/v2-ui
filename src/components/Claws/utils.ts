import BigNumber from 'bignumber.js';
import { INVALID_REASON } from './index';

export interface ValidationParams {
	amount: string | BigNumber;
	maximum: BigNumber;
	minimum?: BigNumber;
}

export function validateAmountBoundaries({
	amount,
	maximum,
	minimum = new BigNumber('0'),
}: ValidationParams): INVALID_REASON | undefined {
	let error: INVALID_REASON | undefined = undefined;

	const input = new BigNumber(amount);
	const amountExceedsBalance = input.gt(maximum);
	const amountIsLessThanMinimum = input.lt(minimum);

	if (amountIsLessThanMinimum) error = INVALID_REASON.UNDER_MINIMUM;
	if (amountExceedsBalance) error = INVALID_REASON.OVER_MAXIMUM;

	console.log({
		input: input.toString(),
		minimum: minimum.toString(),
		maximum: maximum.toString(),
		amountExceedsBalance,
		amountIsLessThanMinimum,
		error,
	});

	return error;
}
