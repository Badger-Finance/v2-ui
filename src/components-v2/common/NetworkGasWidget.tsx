import { Button, ClickAwayListener, Popper } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import { StoreContext } from 'mobx/stores/store-context';
import { observer } from 'mobx-react-lite';
import React, { useContext, useRef } from 'react';

import { getNetworkIconPath } from '../../utils/network-icon';
import NetworkOptions from '../network-selector/NetworkOptions';

const useStyles = makeStyles({
  networkButton: {
    minWidth: 37,
    textAlign: 'center',
    padding: 0,
  },
  selectedNetworkIcon: {
    width: 17,
    height: 17,
  },
  popover: {
    zIndex: 110,
  },
});

const NetworkGasWidget = (): JSX.Element => {
  const classes = useStyles();
  const {
    chain: networkStore,
    uiState: { areNetworkOptionsOpen, openNetworkOptions, closeNetworkOptions },
  } = useContext(StoreContext);
  const ref = useRef<HTMLButtonElement | null>(null);
  return (
    <ClickAwayListener onClickAway={closeNetworkOptions}>
      <div>
        <Button
          ref={ref}
          className={classes.networkButton}
          variant="outlined"
          color="primary"
          onClick={openNetworkOptions}
          aria-label="open network selector"
        >
          <img
            className={classes.selectedNetworkIcon}
            src={getNetworkIconPath(networkStore.network)}
            alt="selected network icon"
            width="17"
            height="17"
          />
        </Button>
        <Popper
          open={areNetworkOptionsOpen}
          className={classes.popover}
          onMouseLeave={closeNetworkOptions}
          anchorEl={ref.current}
          placement="bottom-end"
        >
          <NetworkOptions onSelect={closeNetworkOptions} />
        </Popper>
      </div>
    </ClickAwayListener>
  );
};

export default observer(NetworkGasWidget);
