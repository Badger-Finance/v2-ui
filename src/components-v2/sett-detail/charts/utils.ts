import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { Sett } from '../../../mobx/model/setts/sett';
import { fetchSettChartInformation } from '../../../mobx/utils/apiV2';
import { SettSnapshotGranularity } from '../../../mobx/model/setts/sett-snapshot';
import { SettChartData } from '../../../mobx/model/setts/sett-charts';

dayjs.extend(utc);

export enum SettChartTimeframe {
	'day' = 1,
	'week' = 7,
	'month' = 30,
}

/**
 * Fetches chart information since the provided timeframe until the current date
 * @param sett
 * @param timeframe
 */
export const fetchSettChart = async (sett: Sett, timeframe: SettChartTimeframe): Promise<SettChartData[] | null> => {
	const timeframeDays = 1 * timeframe;
	const isDayTimeFrame = timeframe === SettChartTimeframe.day;

	// if timeframe is just one day then we want the granularity to be hours
	const granularity = isDayTimeFrame ? SettSnapshotGranularity.HOUR : SettSnapshotGranularity.DAY;

	const now = dayjs().utc(); // query until current date
	const from = dayjs(now).subtract(timeframeDays, 'days').utc();

	const fetchedData = await fetchSettChartInformation(sett.vaultToken, from.toDate(), now.toDate(), granularity);

	if (!fetchedData) {
		return null;
	}

	// data needs to be ascending sorted
	return fetchedData
		.sort((a, b) => a.timestamp - b.timestamp)
		.map((d) => ({ ...d, timestamp: new Date(d.timestamp) }));
};
