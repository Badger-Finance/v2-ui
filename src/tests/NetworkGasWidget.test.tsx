import '@testing-library/jest-dom';

import { BadgerAPI } from '@badger-dao/sdk';
import { action } from 'mobx';
import { Chain } from 'mobx/model/network/chain';
import { StoreProvider } from 'mobx/stores/store-context';
import React from 'react';

import NetworkGasWidget from '../components-v2/common/NetworkGasWidget';
import { defaultNetwork, supportedNetworks } from '../config/networks.config';
import GasPricesStore from '../mobx/stores/GasPricesStore';
import store from '../mobx/stores/RootStore';
import { getNetworkIconPath } from '../utils/network-icon';
import { customRender, fireEvent, screen } from './Utils';

const mockGasPrices = {
  rapid: {
    maxPriorityFeePerGas: 2,
    maxFeePerGas: 85.945524054,
  },
  fast: {
    maxPriorityFeePerGas: 2,
    maxFeePerGas: 77.35097164860001,
  },
  standard: {
    maxPriorityFeePerGas: 2,
    maxFeePerGas: 68.7564192432,
  },
  slow: {
    maxPriorityFeePerGas: 2,
    maxFeePerGas: 60.1618668378,
  },
};

function addSpies() {
  jest.spyOn(BadgerAPI.prototype, 'loadVaults').mockReturnValue(Promise.resolve([]));
  jest.spyOn(GasPricesStore.prototype, 'initialized', 'get').mockReturnValue(true);
  jest.spyOn(GasPricesStore.prototype, 'getGasPrices').mockReturnValue(mockGasPrices);
}

describe('NetworkGasWidget', () => {
  const expectedChain = Chain.getChain(defaultNetwork);

  beforeEach(addSpies);

  it('displays network options', () => {
    const { baseElement } = customRender(
      <StoreProvider value={store}>
        <NetworkGasWidget />
      </StoreProvider>,
    );
    fireEvent.click(screen.getByRole('button', { name: 'open network selector' }));
    fireEvent.mouseOver(screen.getByText(expectedChain.name));
    expect(baseElement).toMatchSnapshot();
  });

  it('can select a network', async () => {
    customRender(
      <StoreProvider value={store}>
        <NetworkGasWidget />
      </StoreProvider>,
    );

    const networkIcon = screen.getByAltText('selected network icon');
    fireEvent.click(screen.getByRole('button', { name: 'open network selector' }));
    fireEvent.click(screen.getByText(supportedNetworks[2].name));
    await screen.findByText(supportedNetworks[2].name);
    expect(networkIcon).toHaveAttribute('src', getNetworkIconPath(supportedNetworks[2].network));
  });

  describe('in desktop mode', () => {
    beforeEach(addSpies);

    it('can select gas options', () => {
      const mockSetGasPrice = jest.fn();
      store.chain.setGasPrice = action(mockSetGasPrice);

      customRender(
        <StoreProvider value={store}>
          <NetworkGasWidget />
        </StoreProvider>,
      );

      fireEvent.click(screen.getByRole('button', { name: 'open network selector' }));
      fireEvent.mouseOver(screen.getByText(expectedChain.name));
      fireEvent.click(screen.getByText((mockGasPrices.rapid.maxFeePerGas / 2).toFixed(0)));
      expect(mockSetGasPrice).toHaveBeenNthCalledWith(1, mockGasPrices.rapid);
    });
  });

  // describe('in mobile mode', () => {
  //   beforeEach(addSpies);

  //   it('can select gas options', () => {
  //     const mockSetGasPrice = jest.fn();
  //     store.chain.setGasPrice = action(mockSetGasPrice);
  //     window.matchMedia = createMatchMedia(480);

  //     customRender(
  //       <StoreProvider value={store}>
  //         <NetworkGasWidget />
  //       </StoreProvider>,
  //     );

  //     fireEvent.click(screen.getByRole('button', { name: 'open network selector' }));
  //     fireEvent.click(screen.getAllByRole('button', { name: 'show gas options' })[0]);
  //     fireEvent.click(screen.getByText((mockGasPrices.rapid.maxFeePerGas / 2).toFixed(0)));
  //     expect(mockSetGasPrice).toHaveBeenNthCalledWith(1, mockGasPrices.rapid);
  //   });
  // });
});
