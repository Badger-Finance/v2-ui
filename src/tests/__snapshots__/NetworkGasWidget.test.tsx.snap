// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkGasWidget displays network options 1`] = `
<body>
  <div>
    <div>
      <button
        aria-label="open network selector"
        class="MuiButtonBase-root-33 MuiButton-root-4 MuiButton-outlined-9 makeStyles-networkButton-1 MuiButton-outlinedPrimary-10"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label-5"
        >
          <img
            alt="selected network icon"
            class="makeStyles-selectedNetworkIcon-2"
            src="/assets/icons/ethereum-network.svg"
          />
        </span>
        <span
          class="MuiTouchRipple-root-36"
        />
      </button>
    </div>
  </div>
  <div
    class="makeStyles-popover-3"
    role="tooltip"
    style="position: absolute; top: 0px; left: 0px;"
    x-placement="bottom-end"
  >
    <nav
      class="MuiList-root-50 makeStyles-root-49 makeStyles-root-47 MuiList-subheader-53"
      role="presentation"
    >
      <li
        class="MuiListSubheader-root-56 makeStyles-root-54 MuiListSubheader-sticky-61 makeStyles-sticky-55 MuiListSubheader-gutters-59"
      >
        NETWORK
      </li>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root-33 MuiListItem-root-68 makeStyles-hoveredButton-63 MuiListItem-gutters-75 MuiListItem-button-76 makeStyles-button-67"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiGrid-root-79 MuiGrid-container-80"
        >
          <div
            class="MuiGrid-root-79 MuiGrid-container-80 MuiGrid-item-81 MuiGrid-align-items-xs-center-88 MuiGrid-grid-xs-true-113"
          >
            <div
              class="MuiListItemIcon-root-183 makeStyles-root-182"
            >
              <img
                alt="Ethereum icon"
                class="makeStyles-networkListIcon-62"
                src="/assets/icons/ethereum-network.svg"
              />
            </div>
            <div
              class="MuiListItemText-root-186"
            >
              <span
                class="MuiTypography-root-192 MuiListItemText-primary-190 makeStyles-primary-185 MuiTypography-body1-194 MuiTypography-displayBlock-221"
              >
                Ethereum
              </span>
            </div>
          </div>
          <div
            class="MuiGrid-root-79 MuiGrid-item-81 MuiGrid-grid-xs-auto-112"
          >
            <button
              aria-label="show gas options"
              class="MuiButtonBase-root-33 MuiIconButton-root-222 makeStyles-gasButton-66"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label-230"
              >
                <img
                  alt="display gas options icon"
                  src="/assets/icons/network-selector-arrow.svg"
                />
              </span>
              <span
                class="MuiTouchRipple-root-36"
              />
            </button>
          </div>
        </div>
        <span
          class="MuiTouchRipple-root-36"
        />
      </div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root-33 MuiListItem-root-68 MuiListItem-gutters-75 MuiListItem-button-76 makeStyles-button-67"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiGrid-root-79 MuiGrid-container-80"
        >
          <div
            class="MuiGrid-root-79 MuiGrid-container-80 MuiGrid-item-81 MuiGrid-align-items-xs-center-88 MuiGrid-grid-xs-true-113"
          >
            <div
              class="MuiListItemIcon-root-183 makeStyles-root-182"
            >
              <img
                alt="Local icon"
                class="makeStyles-networkListIcon-62"
                src="/assets/icons/ethereum-network.svg"
              />
            </div>
            <div
              class="MuiListItemText-root-186"
            >
              <span
                class="MuiTypography-root-192 MuiListItemText-primary-190 makeStyles-primary-185 MuiTypography-body1-194 MuiTypography-displayBlock-221"
              >
                Local
              </span>
            </div>
          </div>
          <div
            class="MuiGrid-root-79 MuiGrid-item-81 MuiGrid-grid-xs-auto-112"
          >
            <button
              aria-label="show gas options"
              class="MuiButtonBase-root-33 MuiIconButton-root-222 makeStyles-gasButton-66"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label-230"
              >
                <img
                  alt="display gas options icon"
                  src="/assets/icons/network-selector-arrow.svg"
                />
              </span>
              <span
                class="MuiTouchRipple-root-36"
              />
            </button>
          </div>
        </div>
        <span
          class="MuiTouchRipple-root-36"
        />
      </div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root-33 MuiListItem-root-68 MuiListItem-gutters-75 MuiListItem-button-76 makeStyles-button-67"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiGrid-root-79 MuiGrid-container-80"
        >
          <div
            class="MuiGrid-root-79 MuiGrid-container-80 MuiGrid-item-81 MuiGrid-align-items-xs-center-88 MuiGrid-grid-xs-true-113"
          >
            <div
              class="MuiListItemIcon-root-183 makeStyles-root-182"
            >
              <img
                alt="Polygon icon"
                class="makeStyles-networkListIcon-62"
                src="/assets/icons/matic-network.svg"
              />
            </div>
            <div
              class="MuiListItemText-root-186"
            >
              <span
                class="MuiTypography-root-192 MuiListItemText-primary-190 makeStyles-primary-185 MuiTypography-body1-194 MuiTypography-displayBlock-221"
              >
                Polygon
              </span>
            </div>
          </div>
          <div
            class="MuiGrid-root-79 MuiGrid-item-81 MuiGrid-grid-xs-auto-112"
          >
            <button
              aria-label="show gas options"
              class="MuiButtonBase-root-33 MuiIconButton-root-222 makeStyles-gasButton-66"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label-230"
              >
                <img
                  alt="display gas options icon"
                  src="/assets/icons/network-selector-arrow.svg"
                />
              </span>
              <span
                class="MuiTouchRipple-root-36"
              />
            </button>
          </div>
        </div>
        <span
          class="MuiTouchRipple-root-36"
        />
      </div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root-33 MuiListItem-root-68 MuiListItem-gutters-75 MuiListItem-button-76 makeStyles-button-67"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiGrid-root-79 MuiGrid-container-80"
        >
          <div
            class="MuiGrid-root-79 MuiGrid-container-80 MuiGrid-item-81 MuiGrid-align-items-xs-center-88 MuiGrid-grid-xs-true-113"
          >
            <div
              class="MuiListItemIcon-root-183 makeStyles-root-182"
            >
              <img
                alt="Binance Smart Chain icon"
                class="makeStyles-networkListIcon-62"
                src="/assets/icons/bsc-network.svg"
              />
            </div>
            <div
              class="MuiListItemText-root-186"
            >
              <span
                class="MuiTypography-root-192 MuiListItemText-primary-190 makeStyles-primary-185 MuiTypography-body1-194 MuiTypography-displayBlock-221"
              >
                Binance Smart Chain
              </span>
            </div>
          </div>
          <div
            class="MuiGrid-root-79 MuiGrid-item-81 MuiGrid-grid-xs-auto-112"
          >
            <button
              aria-label="show gas options"
              class="MuiButtonBase-root-33 MuiIconButton-root-222 makeStyles-gasButton-66"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label-230"
              >
                <img
                  alt="display gas options icon"
                  src="/assets/icons/network-selector-arrow.svg"
                />
              </span>
              <span
                class="MuiTouchRipple-root-36"
              />
            </button>
          </div>
        </div>
        <span
          class="MuiTouchRipple-root-36"
        />
      </div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root-33 MuiListItem-root-68 MuiListItem-gutters-75 MuiListItem-button-76 makeStyles-button-67"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiGrid-root-79 MuiGrid-container-80"
        >
          <div
            class="MuiGrid-root-79 MuiGrid-container-80 MuiGrid-item-81 MuiGrid-align-items-xs-center-88 MuiGrid-grid-xs-true-113"
          >
            <div
              class="MuiListItemIcon-root-183 makeStyles-root-182"
            >
              <img
                alt="Arbitrum icon"
                class="makeStyles-networkListIcon-62"
                src="/assets/icons/arbitrum-network.svg"
              />
            </div>
            <div
              class="MuiListItemText-root-186"
            >
              <span
                class="MuiTypography-root-192 MuiListItemText-primary-190 makeStyles-primary-185 MuiTypography-body1-194 MuiTypography-displayBlock-221"
              >
                Arbitrum
              </span>
            </div>
          </div>
          <div
            class="MuiGrid-root-79 MuiGrid-item-81 MuiGrid-grid-xs-auto-112"
          >
            <button
              aria-label="show gas options"
              class="MuiButtonBase-root-33 MuiIconButton-root-222 makeStyles-gasButton-66"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label-230"
              >
                <img
                  alt="display gas options icon"
                  src="/assets/icons/network-selector-arrow.svg"
                />
              </span>
              <span
                class="MuiTouchRipple-root-36"
              />
            </button>
          </div>
        </div>
        <span
          class="MuiTouchRipple-root-36"
        />
      </div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root-33 MuiListItem-root-68 MuiListItem-gutters-75 MuiListItem-button-76 makeStyles-button-67"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiGrid-root-79 MuiGrid-container-80"
        >
          <div
            class="MuiGrid-root-79 MuiGrid-container-80 MuiGrid-item-81 MuiGrid-align-items-xs-center-88 MuiGrid-grid-xs-true-113"
          >
            <div
              class="MuiListItemIcon-root-183 makeStyles-root-182"
            >
              <img
                alt="Fantom icon"
                class="makeStyles-networkListIcon-62"
                src="/assets/icons/fantom-network.svg"
              />
            </div>
            <div
              class="MuiListItemText-root-186"
            >
              <span
                class="MuiTypography-root-192 MuiListItemText-primary-190 makeStyles-primary-185 MuiTypography-body1-194 MuiTypography-displayBlock-221"
              >
                Fantom
              </span>
            </div>
          </div>
          <div
            class="MuiGrid-root-79 MuiGrid-item-81 MuiGrid-grid-xs-auto-112"
          >
            <button
              aria-label="show gas options"
              class="MuiButtonBase-root-33 MuiIconButton-root-222 makeStyles-gasButton-66"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label-230"
              >
                <img
                  alt="display gas options icon"
                  src="/assets/icons/network-selector-arrow.svg"
                />
              </span>
              <span
                class="MuiTouchRipple-root-36"
              />
            </button>
          </div>
        </div>
        <span
          class="MuiTouchRipple-root-36"
        />
      </div>
    </nav>
  </div>
  <div
    class="makeStyles-popper-65"
    role="tooltip"
    style="position: absolute; top: 0px; left: 0px;"
    x-placement="right-start"
  >
    <div
      class="MuiPaper-root-232 MuiPaper-elevation1-236 MuiPaper-rounded-233"
    >
      <nav
        class="MuiList-root-50 makeStyles-root-49 makeStyles-root-231 MuiList-padding-51"
      >
        <div
          aria-disabled="false"
          class="MuiButtonBase-root-33 MuiListItem-root-68 MuiListItem-gutters-75 MuiListItem-button-76 makeStyles-button-67"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root-186"
          >
            <span
              class="MuiTypography-root-192 MuiListItemText-primary-190 makeStyles-primary-185 MuiTypography-body1-194 MuiTypography-displayBlock-221"
            >
               
              43
            </span>
          </div>
          <span
            class="MuiTouchRipple-root-36"
          />
        </div>
        <div
          aria-disabled="false"
          class="MuiButtonBase-root-33 MuiListItem-root-68 MuiListItem-gutters-75 MuiListItem-button-76 makeStyles-button-67"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root-186"
          >
            <span
              class="MuiTypography-root-192 MuiListItemText-primary-190 makeStyles-primary-185 MuiTypography-body1-194 MuiTypography-displayBlock-221"
            >
               
              39
            </span>
          </div>
          <span
            class="MuiTouchRipple-root-36"
          />
        </div>
        <div
          aria-disabled="false"
          class="MuiButtonBase-root-33 MuiListItem-root-68 MuiListItem-gutters-75 MuiListItem-button-76 makeStyles-button-67"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root-186"
          >
            <span
              class="MuiTypography-root-192 MuiListItemText-primary-190 makeStyles-primary-185 MuiTypography-body1-194 MuiTypography-displayBlock-221"
            >
               
              34
            </span>
          </div>
          <span
            class="MuiTouchRipple-root-36"
          />
        </div>
        <div
          aria-disabled="false"
          class="MuiButtonBase-root-33 MuiListItem-root-68 MuiListItem-gutters-75 MuiListItem-button-76 makeStyles-button-67"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root-186"
          >
            <span
              class="MuiTypography-root-192 MuiListItemText-primary-190 makeStyles-primary-185 MuiTypography-body1-194 MuiTypography-displayBlock-221"
            >
               
              30
            </span>
          </div>
          <span
            class="MuiTouchRipple-root-36"
          />
        </div>
      </nav>
    </div>
  </div>
</body>
`;
